### PR TITLE
Add `is_ok` and `is_err` typeguard functions

### LINKED ISSUE #69: PEP647 typing.TypeGuard based is_ok() and is_err() helpers
PEP 647 (https://www.python.org/dev/peps/pep-0647/) makes it possible to write custom type guard functions.

this makes it possible to write shorthand `is_ok(...)` and `is_err(...)` functions that effectively work as shorthands for the longer `isinstance(..., ...)` forms, resulting in denser and more readable code, while still keeping all the benefits of strict type checking:

``` python
def is_ok(result: Result[T, E]) -> TypeGuard[Ok[T]]:
    return isinstance(result, Ok)


def is_err(result: Result[T, E]) -> TypeGuard[Err[E]]:
    return isinstance(result, Err)
```

not all supported python versions support `TypeGuard`, but the widely used `typing_extensions` package (https://github.com/python/typing/blob/master/typing_extensions/README.rst) has a functional backport that works with (at least) mypy. i personally do not see a problem adding a dependency on `typing_extensions`, since it's maintained ‘upstream’, close to python itself, and it will already be pulled in by many other libraries in many real-world applications.

thoughts? opinions? (cc @francium @dbrgn)

### PR DESCRIPTION
**Pull request by mishamsk**
I'd say this is a more performant implementation than #72 as it doesn't use `is instance` checks at all.

Added tests, changelog & took a stab at main readme (although this is up to maintainers, whether this should be advertised as a better approach).

On my laptop one of the mypy tests is failing on 3.9, but that's not related to my change (complains about `|` where `Union` should be)

closes #69

### CODE REVIEW

File: CHANGELOG.md
Comment (francium) [inline CHANGELOG.md:16]: Not sure about the wording here. Perhaps we can leave off the `faster`?

How about,
```suggestion
- `[added]` `is_ok` and `is_err` type guard functions as alternatives to `isinstance` (#69)
```

File: README.rst
Comment (francium) [inline README.rst:80]: I notice the `if/else` pattern, which I'd expect to work in TypeScript doesn't work here. Maybe it's due to pyright not being as advanced as TypeScript's type checker?

```python
x: Result[float, str] = indirection()  # you need an indirection (function that returns a Result), otherwise pyright ignores the type annotation and uses the Ok or Err as a literal

if is_ok(x):
    return x   # Ok[float]
else:
    return x   # Err[str] | Ok[float]     but you'd expect it to resolve to Err[str]
```

Comment (francium) [reply README.rst:80]: TypeScript reference: https://www.typescriptlang.org/play?#code/C4TwDgpgBAogTnKBeKBvKBrAlgOwCYBcUA5BAsVAL4BQokUA8hsmproSQPYYU13QAlCAGcArgBtgLRswA+sBNWoBjTjmFSswgPrcWACjhEhYyQEoiiLTOQA+KHAB02fMiQpi3YktXrN+LDgIZWAsNQMLKBMJKSR7fXQXDlJyKjMfNQ0oAA9jERiWdkDg0LV9MwBuJSwAMyh9LV0MfWyzMzRqKC6ciqgAej6ZakooCHFhaFRO7uzegYU4YaA
```typescript
type Err = { kind: 'err' }
type Ok = { kind: 'ok' }
type Result =  Ok | Err

const is_ok = (r: Result): r is Ok => r.kind === 'ok'

const indirection = (): Result => ({ kind: 'err' })

const x: Result = indirection();

if (is_ok(x)) {
    x; // Ok
} else {
    x; // Err
}
```

File: src/result/result.py
Comment (francium) [inline src/result/result.py:467]: As I understand it, `reveal_type` is a mypy feature, not a valid python function. Perhaps we can just add a comment suggesting the type of `result` is `Ok[int]`?
```suggestion
 >>>     result  # type of result is Ok[int]
```

File: src/result/result.py
Comment (francium) [inline src/result/result.py:469]: similarly here

File: src/result/result.py
Comment (francium) [inline src/result/result.py:455]: and here

File: src/result/result.py
Comment (francium) [inline src/result/result.py:465]: Can we generalize this example to something like,
```
r: Result[int, str] = get_a_result()
if is_ok(r):
    r   # r is of type Ok[int]
elif is_err(r):
    r   # r is of type Err[str]
```

Just so we have a smaller example, with only 1 variable.

File: README.rst
Comment (francium) [reply README.rst:80]: I re-read the discussion over on #69, this was already mentioned here
https://github.com/rustedpy/result/issues/69#issuecomment-1151233994

And there's an upstream discussion in the python/typing repo https://github.com/python/typing/discussions/1013

So we can ignore this for now

File: src/result/__init__.py
Comment (francium) [inline src/result/__init__.py:22]: These technically will conflict with the `Err.is_ok` and `Ok.is_ok` methods.
Might cause some confusion?
But as long as someone is using a type checker with this library, they'll realize those methods aren't type guards and will use the proper `is_ok`/`is_err` type guard functions...hopefully I guess.

**Review by francium — CHANGES_REQUESTED** [review]
Just a few minor things.

Otherwise PR looks great. Thank you for the help.

@wbolster if you have any thought or opinions on this, since you were working in this area prior, feel free to take a look and share your comments.

File: README.rst
Comment (mishamsk) [reply README.rst:80]: yeah, if they write and implement the new PEP, this would just be a matter of doc change. I hope they will

File: src/result/result.py
Comment (mishamsk) [reply src/result/result.py]: applied your version

File: src/result/result.py
Comment (mishamsk) [reply src/result/result.py]: applied your version

File: README.rst
Comment (francium) [inline README.rst:124]: Note to self, small typo here, remove the `.` at the end of `safe`

File: CHANGELOG.md
Comment (francium) [inline CHANGELOG.md:16]: Note to self, small typo here, remove the extra `-` at the start of this line

File: README.rst
Comment (francium) [inline README.rst:126]: Note to self, perhaps make `methods` and `functions` bold to distinguish them since `is_ok` and `is_err` have the same names

**Comment by wbolster** [comment]
sorry for late reply. !72 suffered from some type checker limitations, and this one does not. do i understand that correctly?

> I'd say this is a more performant implementation than #72 as it doesn't use `isinstance()` checks at all.

i think this is misguided. why would `isinstance()` checks be slow? in cpython, these usually boil down to a direct pointer comparison, which is always much faster than a method call on a custom class.

**Comment by mishamsk** [comment]
> sorry for late reply. #72 suffered from some type checker limitations, and this one does not. do i understand that correctly?

unfortunately not, it is all the same. But a new PEP is being [drafted](https://github.com/python/peps/pull/3266) which will "fix" that

> > I'd say this is a more performant implementation than #72 as it doesn't use `isinstance()` checks at all.
> 
> i think this is misguided. why would `isinstance()` checks be slow? in cpython, these usually boil down to a direct pointer comparison, which is always much faster than a method call on a custom class.

Oh, I assumed it should be slower, but apparently I am wrong. Although it may be different in 3.11 with reduced "price" for additional frames and specializing attribute getters, which I'd hope would speed up bound methods getters as well... Thanks for the tip about isinstance checks. Do you happen to have a good link handy that explores how CPython implements them by any chance?

**Comment by wbolster** [comment]
>> i think this is misguided. why would `isinstance()` checks be slow? in cpython, these usually boil down to a direct pointer comparison, which is always much faster than a method call on a custom class.
>
> Thanks for the tip about isinstance checks. Do you happen to have a good link handy that explores how CPython implements them by any chance?

@mishamsk

this should help:

- `isinstance()` (python) corresponds to `PyObject_IsInstance` (c api)
- `PyObject_IsInstance` (https://github.com/python/cpython/blob/v3.11.4/Objects/abstract.c#L2682C1-L2682C20) delegates to `object_recursive_isinstance()`
- `object_recursive_isinstance()` (https://github.com/python/cpython/blob/v3.11.4/Objects/abstract.c#L2617C1-L2617C28) has a fast path using `Py_IS_TYPE` (https://github.com/python/cpython/blob/v3.11.4/Objects/abstract.c#L2620C9-L2620C19)
- `Py_IS_TYPE` is a macro ([docs](https://docs.python.org/3/c-api/structures.html#c.Py_IS_TYPE)) expanding to `Py_TYPE(o) == type`
- `Py_TYPE` ((docs)[https://docs.python.org/3/c-api/structures.html#c.Py_TYPE]) boils down to
`return ob->ob_type` ((source)[https://github.com/python/cpython/blob/v3.11.4/Include/object.h#L149])

… so the simple cases boil down to a bit of pointer chasing and a pointer comparison, which is all very fast.

**Comment by mishamsk** [comment]
@wbolster thanks a lot! I got curious, and run a very dumb test (sorry for formatting, doing it from iPad while on vacation):

```python
import timeit

class T:
    cattr = True

    def __init__(self) -> None:
        self.attr = True

    @property
    def prop(self) -> bool:
        return True

    def ist(self) -> bool:
        return True
    
class T1:
    __slots__ = ("attr",)

    cattr = True

    def __init__(self) -> None:
        self.attr = True

    @property
    def prop(self) -> bool:
        return True

    def ist(self) -> bool:
        return True

t = T()

inst_check = timeit.timeit("isinstance(t, T)", globals=globals())
fn_check = timeit.timeit("t.ist()", globals=globals())
prop_check = timeit.timeit("t.prop", globals=globals())
attr_check = timeit.timeit("t.attr", globals=globals())
cattr_check = timeit.timeit("t.cattr", globals=globals())

print(f"Non-slotted. Isintance: {inst_check}. Method: {fn_check}. Prop: {prop_check}. Inst attr: {attr_check}. Class attr: {cattr_check}")

t = T1()

inst_check = timeit.timeit("isinstance(t, T)", globals=globals())
fn_check = timeit.timeit("t.ist()", globals=globals())
prop_check = timeit.timeit("t.prop", globals=globals())
attr_check = timeit.timeit("t.attr", globals=globals())
cattr_check = timeit.timeit("t.cattr", globals=globals())

print(f"Slotted. Isintance: {inst_check}. Method: {fn_check}. Prop: {prop_check}. Inst attr: {attr_check}. Class attr: {cattr_check}")
```

Which gives for 3.10 & 3.11 (in docker running on synology NAS;-), but we do not care about absolute numbers):
```sh
py310: commands[0]> python isinstance_test.py
Non-slotted. Isintance: 0.16529686748981476. Method: 0.22358105331659317. Prop: 0.20174939185380936. Inst attr: 0.10315892100334167. Class attr: 0.10350674390792847
Slotted. Isintance: 0.2414640188217163. Method: 0.1983325481414795. Prop: 0.20569878816604614. Inst attr: 0.09698593616485596. Class attr: 0.08725540339946747
py310: OK ✔ in 1.99 seconds
py311: commands[0]> python isinstance_test.py
Non-slotted. Isintance: 0.0721818059682846. Method: 0.10906121879816055. Prop: 0.16895851492881775. Inst attr: 0.03669581562280655. Class attr: 0.0829509049654007
Slotted. Isintance: 0.12867401540279388. Method: 0.10731867700815201. Prop: 0.17078260332345963. Inst attr: 0.035740092396736145. Class attr: 0.06810076534748077
  py310: OK (1.99=setup[0.30]+cmd[1.68] seconds)
  py311: OK (1.05=setup[0.01]+cmd[1.04] seconds)
  congratulations :) (3.19 seconds)
```

So, apparently, speedwise the optimal way would be to create an instance attribute (something like `__is_ok__`) and use it, which will be quicker than isinstance checks. But, until the PEP I’ve mentioned is approved, `is_ok`/`is_err` functions would still be more limiting than isinstance checks typing wise.

PS. surprised to see that property access is slower than a method call on py311. I guess some of their call optimisations do not work for descriptors🤷‍♂️

**Pull request merged**
